### ISSUE TITLE
Module execution order can be incorrect when having multiple entrypoints

### ISSUE THREAD

**Issue by sapphi-red** [rollup/rollup#5423]
### Rollup Version

4.12.1

### Operating System (or Browser)

StackBlitz / REPL

### Link To Reproduction

https://rollupjs.org/repl/?version=4.12.1&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU1QyUyMi4lMkYxLmpzJTVDJTIyJTVDbmltcG9ydCUyMCU1QyUyMi4lMkYyLmpzJTVDJTIyJTVDbmltcG9ydCUyMCU1QyUyMi4lMkYzLmpzJTVDJTIyJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyY29uc29sZS5sb2coJzEnKSUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjIxLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmNvbnNvbGUubG9nKCcyJyklMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIyMi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJjb25zb2xlLmxvZygnMycpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMjMuanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwJTVDJTIyLiUyRjEuanMlNUMlMjIlNUNuaW1wb3J0JTIwJTVDJTIyLiUyRjMuanMlNUMlMjIlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJzdWIuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE , https://stackblitz.com/edit/rollup-repro-9wagvo?file=src%2Fmain.js

### Expected Behaviour

```js
// main.js
import './1-hash.js'
import './2-hash.js'
import './3-hash.js'

// sub.js
import './1-hash.js'
import './3-hash.js'

// 1-hash.js
console.log('1');

// 2-hash.js
console.log('2');

// 3-hash.js
console.log('3');
```

### Actual Behaviour

```js
// main.js
import './3-hash.js'

console.log('2');

// sub.js
import './3-hash.js'

// 3-hash.js
console.log('1');

console.log('3');
```

**Comment by lukastaegert** [comment]
Yes, I am aware of this and it is indeed a duplicate. One problem is that fixing this would involve creating quite a few more chunks, and we already have a lot of chunks in complicated projects. So I wonder if this "strict execution order" should be an opt-in or an opt-out for users via an option. What would you think?
Another thought would be to track which modules have side effects and only ensure correct execution order between modules with side effects. Then maybe we could just enable it for everyone and hope that people do not have too many side effects in their modules.

**Comment by sapphi-red** [comment]
In Vite, we transform each HTML into a module with `<link rel=stylesheet href=/foo.css />` transformed to `import './foo.css'`. We set `moduleSideEffects: 'no-treeshake'` to CSS files and HTML files. Then, we inject the CSS tag rely on the generated chunk order. So in this specific case, opt-in for each module would work for Vite.

I guess in most cases, users would want to apply it to some modules like modules injecting polyfills.

**Comment by lukastaegert** [comment]
For polyfills, there is usually not a problem if you ensure that all static entries first import the polyfill. See also https://rollupjs.org/faqs/#how-do-i-add-polyfills-to-a-rollup-bundle . But it would still be nice to fix this properly.

**Comment by sapphi-red** [comment]
Ah, that's true.

If rollup goes with the latter way (= track which modules have side effects), Vite would need to inject a dummy statement that has side effects and remove that later in `renderChunk`. So it'd be nice if there's a way to opt-in per module even with the latter way.

---

Maybe starting with an new option that opt-in the behavior is good. Then, I guess we can know how much the number of chunks will increase. Also we can try if `experimentalMinChunkSize` can eliminate the increase.

**Comment by lukastaegert** [comment]
Setting `moduleSideEffects: 'no-treeshake'` is always considered one hell of a side effect, so those would be properly ordered 😉
